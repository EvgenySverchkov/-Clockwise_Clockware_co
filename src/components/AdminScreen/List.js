import React from 'react';

export default function CreateList({ListItem, dataArr, style}){
  function CreateHeaders({dataArr}){
    if(dataArr.length===0){
      return (<tr><th>List is empty</th></tr>)
    }
    let keysArr = Object.keys(dataArr[0]);
    return (
      <tr>
        {
          keysArr.map(item=>{
            let value = item;
            if(typeof item === 'string'){
              let upperChar = item.match(/[A-Z]/);
              let strCopy = item;
              if(upperChar){
                let idx = upperChar.index;
                strCopy = item.slice(0, idx) + " " + item[idx].toLocaleLowerCase() + item.slice(idx+1);
              }
              value = strCopy[0].toUpperCase() + strCopy.slice(1);
            }
            return (<th key={item}>{value}</th>);
          })
        }
        <th></th>
      </tr>
    )
  }
  return (
    <table className="table table-dark" style={style}>
      <thead>
        <CreateHeaders dataArr={dataArr}/>
      </thead>
      <tbody>
        {
          dataArr.map((item)=>(
            <tr key={item.id+1}>
              <ListItem infoObj = {item}/>
            </tr>
          ))
        }
      </tbody>
    </table>
  );
}
