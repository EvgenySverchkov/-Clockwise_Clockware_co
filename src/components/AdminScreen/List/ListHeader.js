import React from 'react';

export default function ListHeader({templObj}){
  let keysArr = Object.keys(templObj||{});
  if(keysArr.length===0){
    return (<tr><th>List is empty</th></tr>)
  }
  return (
    <tr>
      {
        keysArr.map(item=>{
          let value = item;
          if(typeof item === 'string'){
            let upperChar = item.match(/[A-Z]/);
            let strCopy = item;
            if(upperChar){
              let idx = upperChar.index;
              strCopy = item.slice(0, idx) + " " + item[idx].toLocaleLowerCase() + item.slice(idx+1);
            }
            value = strCopy[0].toUpperCase() + strCopy.slice(1);
          }
          return (<th key={item} className="text-center" style={{wordWrap: "break-word"}}>{value}</th>);
        })
      }
      <th></th>
    </tr>
  )
}
