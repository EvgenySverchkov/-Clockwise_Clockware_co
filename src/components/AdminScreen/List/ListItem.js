import React from 'react';
import PropTypes from 'prop-types';

function ListItem({infoObj}){
  let keysArr = Object.keys(infoObj);
  return (
    <>
      {
        keysArr.map((item)=><td className="text-center"
                                style={{wordWrap: "break-word"}}
                                key={item}>{infoObj[item]}</td>)
      }
    </>
  );
}

ListItem.propTypes = {
  infoObj: PropTypes.object
}

export default ListItem;
