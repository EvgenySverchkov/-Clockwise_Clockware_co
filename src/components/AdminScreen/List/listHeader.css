th{
    position: sticky; 
    top: -1px;
    background: #343a40 ;
    z-index: 1;
}