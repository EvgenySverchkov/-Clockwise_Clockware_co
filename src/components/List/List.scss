.table-responsive {
  overflow-y: auto;
  height: 80vh;
}
