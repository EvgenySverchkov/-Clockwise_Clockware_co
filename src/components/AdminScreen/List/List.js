import React, { useEffect } from "react";
import PropTypes from "prop-types";

import OptionsBtns from "../OptionsBtnGroup";
import ListItem from "./ListItem";
import ListHeader from "./ListHeader";

function CreateList({
  dataArr,
  deleteAction,
  mainRows,
  getData,
  history,
  listIsLoad,
}) {
  useEffect(
    function () {
      getData();
    },
    [history.location]
  );
  return (
    <div
      className="table-responsive"
      style={{ overflowY: "auto", height: "80vh" }}
    >
      <table className="table table-dark">
        <thead>
          {listIsLoad ? (
            <tr>
              <th>Loading...</th>
            </tr>
          ) : dataArr.length === 0 ? (
            <tr>
              <th>List is empty</th>
            </tr>
          ) : (
            <ListHeader templArr={mainRows} />
          )}
        </thead>
        <tbody>
          {dataArr.map((item) => (
            <tr key={item.id}>
              <ListItem infoObj={item} mainRows={mainRows} />
              <OptionsBtns deleteMasterById={deleteAction} itemObj={item} />
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}
CreateList.propTypes = {
  dataArr: PropTypes.array,
  style: PropTypes.object,
  deleteAction: PropTypes.func,
};

export default CreateList;
