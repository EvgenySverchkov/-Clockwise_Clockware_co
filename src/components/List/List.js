import React, { useEffect } from "react";
import PropTypes from "prop-types";

import OptionsBtns from "../OptionsBtnGroup";
import ListItem from "./ListItem";
import ListHeader from "./ListHeader";
import ShowFullUserOrderInfo from "../OptionsBtnGroup/ShowFullUserOrderInfo";

import "./List.scss";

function CreateList({
  dataArr,
  deleteAction,
  mainRows,
  getData,
  history,
  listIsLoad,
}) {
  useEffect(
    function () {
      getData();
    },
    [history.location]
  );
  return (
    <div className="table-responsive">
      <table className="table table-dark">
        <thead>
          {listIsLoad ? (
            <tr>
              <th>Loading...</th>
            </tr>
          ) : dataArr.length === 0 ? (
            <tr>
              <th>List is empty</th>
            </tr>
          ) : (
            <ListHeader templArr={mainRows} />
          )}
        </thead>
        <tbody>
          {dataArr.map((item) => (
            <tr key={item.id}>
              <ListItem infoObj={item} mainRows={mainRows} />
              {deleteAction ? (
                <OptionsBtns deleteMasterById={deleteAction} itemObj={item} />
              ) : (
                <ShowFullUserOrderInfo itemObj={item} />
              )}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}
CreateList.propTypes = {
  dataArr: PropTypes.array,
  style: PropTypes.object,
  deleteAction: PropTypes.func,
};

export default CreateList;
