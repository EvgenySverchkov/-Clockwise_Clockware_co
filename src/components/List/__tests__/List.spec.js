import React from "react";
import {create} from "react-test-renderer";
import {MemoryRouter} from "react-router-dom";
import {Provider} from "react-redux";
import configMockStore from "redux-mock-store";

import List from "../List";

const mockStore = configMockStore();

describe("Test of <List />", ()=>{
    it("View when in data array 2 items", ()=>{
        const store = mockStore({});
        const mockProps = {
            dataArr: [{
                id: 1,
                rating: 5,
                name: "Petrovich",
                towns: ""
            },{
                id: 2,
                rating: 5,
                name: "Alexeevich",
                towns: ""
            }],
            deleteAction: jest.fn(),
            mainRows: ["name", "rating"],
            getData: jest.fn(),
            history: {location: "/"},
            listIsLoad: false,
        }
        const component = create(
            <MemoryRouter initialEntries={["/"]}>
                <Provider store = {store}>
                    <List {...mockProps}/>
                </Provider>
            </MemoryRouter>
        )
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("View when data array is empty", ()=>{
        const store = mockStore({});
        const mockProps = {
            dataArr: [],
            deleteAction: jest.fn(),
            mainRows: ["name", "rating"],
            getData: jest.fn(),
            history: {location: "/"},
            listIsLoad: false,
        }
        const component = create(
            <MemoryRouter initialEntries={["/"]}>
                <Provider store = {store}>
                    <List {...mockProps}/>
                </Provider>
            </MemoryRouter>
        )
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("View when data array is loading", ()=>{
        const store = mockStore({});
        const mockProps = {
            dataArr: [],
            deleteAction: jest.fn(),
            mainRows: ["name", "rating"],
            getData: jest.fn(),
            history: {location: "/"},
            listIsLoad: true,
        }
        const component = create(
            <MemoryRouter initialEntries={["/"]}>
                <Provider store = {store}>
                    <List {...mockProps}/>
                </Provider>
            </MemoryRouter>
        )
        expect(component.toJSON()).toMatchSnapshot();
    });
});