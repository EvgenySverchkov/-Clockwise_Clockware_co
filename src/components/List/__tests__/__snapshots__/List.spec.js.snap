// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <List /> View when data array is empty 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-dark"
  >
    <thead>
      <tr>
        <th>
          List is empty
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`Test of <List /> View when data array is loading 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-dark"
  >
    <thead>
      <tr>
        <th>
          Loading...
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`Test of <List /> View when in data array 2 items 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-dark"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Rating
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style={
            Object {
              "wordWrap": "break-word",
            }
          }
        >
          Petrovich
        </td>
        <td
          style={
            Object {
              "wordWrap": "break-word",
            }
          }
        >
          5
        </td>
        <td
          className="text-center"
        >
          <div
            className="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              className="btn btn-secondary dropdown-toggle"
              data-toggle="dropdown"
              id="dropdownMenuButton"
              type="button"
            >
              Options
            </button>
            <div
              aria-labelledby="dropdownMenuLink"
              className="dropdown-menu"
            >
              <button
                className="dropdown-item deleteItemBtn"
                onClick={[Function]}
              >
                Delete
              </button>
              <a
                className="dropdown-item"
                href="/1"
                onClick={[Function]}
              >
                Edit
              </a>
              <button
                className="dropdown-item"
                data-target="#fullInfoModal"
                data-toggle="modal"
                onClick={[Function]}
                type="button"
              >
                Show full info
              </button>
            </div>
          </div>
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "wordWrap": "break-word",
            }
          }
        >
          Alexeevich
        </td>
        <td
          style={
            Object {
              "wordWrap": "break-word",
            }
          }
        >
          5
        </td>
        <td
          className="text-center"
        >
          <div
            className="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              className="btn btn-secondary dropdown-toggle"
              data-toggle="dropdown"
              id="dropdownMenuButton"
              type="button"
            >
              Options
            </button>
            <div
              aria-labelledby="dropdownMenuLink"
              className="dropdown-menu"
            >
              <button
                className="dropdown-item deleteItemBtn"
                onClick={[Function]}
              >
                Delete
              </button>
              <a
                className="dropdown-item"
                href="/2"
                onClick={[Function]}
              >
                Edit
              </a>
              <button
                className="dropdown-item"
                data-target="#fullInfoModal"
                data-toggle="modal"
                onClick={[Function]}
                type="button"
              >
                Show full info
              </button>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
